---
Title: RRBS Pipeline
Author: Emily Isenhart
Date: 5/10/24
---

Run with cmd+optn+enter
```bash
# set up conda 
eval "$(/util/common/python/py38/anaconda3-2020.11/bin/conda shell.bash hook)"

###############################################################################
# run trim galore
## create a temporary environment to load everything in
## this avoid permissions issues with writing to the common environment 
conda create --prefix /tmp/test-env python=2.7
conda activate /tmp/test-env
### conda deactivate (use when done to end)

## load Bioconda
## documentation: https://bioconda.github.io/
conda config --add channels defaults
conda config --add channels bioconda
conda config --add channels conda-forge
conda config --set channel_priority strict

## install trim galore to temporary environment 
conda install trim-galore

## run trim galore to remove illumina adapter sequences for paired end reads
mkdir Trimmed 
N=4
(for read1 in RQ025150-Ohm/*/*R1_001.fastq.gz; do \
((i=i%N)); ((i++==0)) && wait \
read2=$(echo $read1| sed 's/R1_001.fastq.gz/R3_001.fastq.gz/'); \
trim_galore --paired -a AGATCGGAAGAGC -a2 AAATCAAAAAAAC $read1 $read2 \
-o Trimmed & done)

## delete directory
### leaving delete commented out for my own safety
find RQ025150-Ohm #-delete

###############################################################################
# trim diversity adapters
## pull the script from NuGen's git repo 
## https://github.com/nugentechnologies/NuMetRRBS
curl -o Scripts/trimRRBSdiversity.py \
https://raw.githubusercontent.com/nugentechnologies/NuMetRRBS/master/\
trimRRBSdiversityAdaptCustomers.py

## run NuGen diversity trimming script
for read1 in Trimmed/*R1_001_val_1.fq.gz; do \
read2=$(echo $read1| sed 's/R1_001_val_1.fq.gz/R3_001_val_2.fq.gz/'); \
python Scripts/trimRRBSdiversity.py -1 $read1 -2 $read2 -o Trimmed & done

## delete files
### leaving delete commented out for my own safety
find Trimmed/*R1_001_val_1.fq.gz #-delete
find Trimmed/*R3_001_val_2.fq.gz #-delete
find Scripts/trimRRBSdiversity.py -delete

###############################################################################
# run fastqc
mkdir Fastqc
module load fastqc
fastqc -o Fastqc -t 4 Trimmed/*trimmed.fq.gz

###############################################################################
# run bismark
## align pared-end read to the genome
mkdir Aligned
conda install bismark
module load bimark

## run alignment 
N=4
(for read1 in Trimmed/*R1_001_val_1.fq_trimmed.fq.gz; do \
((i=i%N)); ((i++==0)) && wait \
read2=$(echo $read1| \
sed 's/R1_001_val_1.fq_trimmed.fq.gz/R3_001_val_2.fq_trimmed.fq.gz/'); \
bismark --bowtie2 --gzip --sam /projects/rpci/shared/reference/GRCh38_bismark/ \
-1 $read1 -2 $read2 & done)

## for whatever reason, I can't get these files to output where I want them 
## so I'm just going to move them 
mv *.gz \
Aligned

###############################################################################
# run de-duplication 
## this can be done for RRBS using NuGen becuase they introduce a unique 
## molecular tag for identifying duplicated reads rather than the 
## typical deduplication technique of identifying overlapping genomic regions 
## (NOT for use in targeted assay like RRBS)

## pull NuDup script from NuGen
### May 13th- remember to document the details of how Nudup runs on paired-end 
### reads in the ReadMe for this git 
curl -o Scripts nudup.py \
https://raw.githubusercontent.com/tecangenomics/nudup/master/nudup.py
### run script 

## delete files
find Scripts/Scripts nudup.py -delete


###############################################################################
# run bismark-rDNA?

###############################################################################
# run bismark-methylex

###############################################################################

```
